Initial Rails Commands Infrastructure

kaspth opened the pull request:
Adds the initial infrastructure needed to run our internal commands with Thor. It unifies the regular app commands and engine commands into one.

I've left the old command files in place in rails/commands and rails/engine/commands to slim the diff and ease review. I'll remove them before the final merge.

In another PR we can add command dependencies and port our Rake commands over.

Later, we can figure out how to let users add their own commands. It's mostly there; there's lookup code, the app commands just needs to be on the lookup path.

kaspth committed:
Initial command structure.

kaspth committed:
Run engine commands through command infrastructure.

prathamesh-sonpatki commented:
@kaspth Any specific reason why we have each command in its own folder like `/commands/rake/rake_command.rb`?

eileencodes commented on railties/lib/rails/command.rb:
```diff
@@ -0,0 +1,96 @@
+module Rails
+  module Command
+    autoload :Behavior, "rails/command/behavior"
+    autoload :Base, "rails/command/base"
```
Autoload isn't finding these without the path?

kaspth replied:
Oh, it works relatively? Thanks, I'll try it!

kaspth replied:
Ruby's standard autoload wants a file mapping as well. ¯\_(ツ)_/¯

eileencodes replied:
Yea not sure how exactly that works, some places in Rails we're using it without paths, and then in other places we are using paths.  As you said ¯_(ツ)_/¯

fxn replied:
Because those ones use ActiveSupport::Autoload (on my phone, from memory). This is the standard in the code base, conventional file layout and no paths.

kaspth replied:
Oh right, I'll pull that module in and save those paths 😁

eileencodes commented on railties/lib/rails/command.rb:
```diff
@@ -0,0 +1,96 @@
+        @hidden_commands ||= []
+      end
+
+      def environment
```
Should this be no doc'd or documented? Looks like probably a nodoc.

kaspth replied:
Yeah, executing a command in a given environment is probably a Rails internal thing. On the other hand having every command except a `--environment` could be neat (which this method could help assign the default for).

kaspth replied:
Nodoc'ing for now 👍

eileencodes commented on railties/lib/rails/command.rb:
```diff
@@ -0,0 +1,96 @@
+        groups.transform_values! { |commands| commands.flat_map(&:printing_commands).sort }
+
+        rails = groups.delete("rails")
+        [[ "rails", rails ]] + groups.sort.to_a
```
Why delete `"rails"` and then add it back?

kaspth replied:
To make Rails' built in commands show up first in help output, e.g.:

```
Rails:
  console
  dbconsole
  destroy
  generate
  new
  runner
  server
  test
  version

Rake:
  about
  app:template
  app:update
  assets:clean[keep]
  assets:clobber
```

RichardBlair replied:
In that case it might make more sense to remove rails from the groups, group the commands, transform, and then prepend rails?

RichardBlair replied:
Sorry, just realized this was merged. I can just follow up with a PR.

eileencodes commented on railties/lib/rails/command/base.rb:
```diff
@@ -0,0 +1,135 @@
+
+        # Tries to get the description from a USAGE file one folder above the command
+        # root.
+        def desc(usage = nil, description = nil)
```
Did you call this `desc` instead of `description` because someone else defines description already? Abbreviated this way I read it was `descending`.

kaspth replied:
The `desc` is from Thor: https://github.com/erikhuda/thor/blob/06dade7c9567edccc4da2f0d6fcf4c2d00066e3d/lib/thor.rb#L54

We could add an alias for this description, which I agree would be more... descriptive 🤓

eileencodes commented on railties/lib/rails/command/environment_argument.rb:
```diff
@@ -0,0 +1,32 @@
+          if available_environments.include? env
+            env
+          else
+            %w(production development test).detect { |e| e =~ /^#{env}/ } || env
```
I think we prefer spaces around arrays like this: `%w( production development test )...`

kaspth commented:
@prathamesh-sonpatki mostly to make it easier to add a USAGE file or maybe other tertiary files later on.

kaspth commented:
Note that this also removes the command whitelist restrictions on the engine commands. I'll need to check the added commands more to see if they work out of the box.

E.g. it would be neat if the server command could find out if it was in an engine and then just boot `test/dummy`.

kaspth committed:
Simplify auxiliary command files directory finding.

Still vulnerable to different file structures. We likely want something
more robust when we tackle in app commands.

kaspth committed:
Per Dr. Eileen's orders :)

Prescribed some review fixes for myself!

eileencodes commented on railties/lib/rails/commands/server/server_command.rb:
```diff
@@ -0,0 +1,159 @@
+    end
+
+    private
+
```
✂️

maclover7 commented:
👍 this is looking really good! Quick question just to confirm, commands override the `#perform` method to run the command, right?

aliAsadi92 commented:
👍

kaspth commented:
@maclover7 not sure what you mean? The instance level `perform` is just added as an alias for the actual command name, e.g. the test command would want an instance level `test` method. But I found that to be too much repetition for our internal common case of one command per file.

So yes, right now they're one command per file. But it shouldn't have to be that way, we just don't have to tackle it upfront and let the infrastructure as we solve more use cases. This is a starting point 😁

kaspth committed:
Remove the old command files.

Wash out your old! These adhoc scripts are replaced by the new
commands.

kaspth commented on railties/lib/rails/command/behavior.rb:
```diff
@@ -0,0 +1,123 @@
+
+          # This code is based directly on the Text gem implementation.
+          # Copyright (c) 2006-2013 Paul Battley, Michael Neumann, Tim Fletcher.
+          #
```
Pssst @matthewd

kaspth commented:
Removed the old command files 😁

kaspth committed:
Fix console tests.

rafaelfranca commented on railties/lib/rails/command/behavior.rb:
```diff
@@ -0,0 +1,123 @@
+          # This code is based directly on the Text gem implementation.
+          # Copyright (c) 2006-2013 Paul Battley, Michael Neumann, Tim Fletcher.
+          #
+          # Returns a value representing the "cost" of transforming str1 into str2
```
`.` in the end.

rafaelfranca commented on railties/lib/rails/commands/server/server_command.rb:
```diff
@@ -88,9 +90,7 @@ def middleware
-      super.merge(
-        Port:               ENV.fetch("PORT", 3000).to_i,
-        Host:               ENV.fetch("HOST", "localhost").dup,
+      super.merge(        Port:               ENV.fetch("PORT", 3000).to_i,
```
Why Host was removed? and also the style is wrong here.

kaspth replied:
A fault from a rebase I think. Fixed, thanks!

rafaelfranca commented on railties/lib/rails/test_unit/minitest_plugin.rb:
```diff
@@ -1,4 +1,5 @@
 require "active_support/core_ext/module/attribute_accessors"
+require "active_support/core_ext/hash/keys"
```
Why is this here?

kaspth replied:
Some leftover from an earlier change.

rafaelfranca commented:
Tests are also broken.

kaspth committed:
Apply Rafaels review fixes.

kaspth committed:
Fix dbconsole tests.

kaspth committed:
Fix server command tests.

kaspth commented:
Got the build back to green again. @rafaelfranca since you requested the last changes, what do you say?

andremedeiros commented on railties/lib/rails/command/behavior.rb:
```diff
@@ -0,0 +1,123 @@
+          # Copyright (c) 2006-2013 Paul Battley, Michael Neumann, Tim Fletcher.
+          #
+          # Returns a value representing the "cost" of transforming str1 into str2.
+          def levenshtein_distance(str1, str2)
```
I could totally see a use for this in the ActiveSupport string extensions.

kaspth replied:
I'm not sure we want to make this a public API. Maybe as `edit_distance_to('other string')`, but when would you want the distance directly versus just sort an array by it?

E.g. this might be more suitable:

```ruby
%w( some words with a distance ).sort_by_edit_distances_to('this string')
```

andremedeiros replied:
Sounds like the former would totally back up the latter here, right? Or am I missing something?

kaspth replied:
Yes, but I'm trying if and how we'd want to expose this as an API :)

kaspth replied:
Spoke with the team and we're not gonna provide this as a public API. But maybe you can try the did_you_mean gem? It's bundled with Ruby 2.3 and also includes this code: https://github.com/yuki24/did_you_mean/blob/c22ef694f411ed4f656a029630847129c78c4cb1/lib/did_you_mean/levenshtein.rb

kaspth committed:
Apply test command changes.
